File: question/neetcode/binary_tree/maximum-depth-of-binary-tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    """
            1
        2       3
    4

    1: max(1+fun(2), 1+fun(3))
    2: max(1+fun(4), 1+fun(null))

    fun(null) return 0
    """

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        return 1+max(self.maxDepth(root.left), self.maxDepth(root.right))

File: question/neetcode/binary_tree/maximum-depth-of-binary-tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void dfs(TreeNode* node, int c, int& h){
        if(not node) return;

        h = max(c,h);
        dfs(node->left, c+1, h);
        dfs(node->right, c+1, h);
    }

    int maxDepth(TreeNode* root) {
        int m = 0;
        dfs(root, 1, m);

        return m;
    }
};